tests/test_otel.py

```diff
@@ -0,0 +1,318 @@
+            otel_span_attr http.request.completion
+                $request_completion;
+            otel_span_attr http.response.header.content.type
+                $sent_http_content_type;
+            otel_span_attr http.request $request;
+            return 200 "OK";
+        }
+
+        location /vars {
+            otel_trace_context {{ context or "ignore" }};
+            add_header "X-Otel-Trace-Id" $otel_trace_id;
+            add_header "X-Otel-Span-Id" $otel_span_id;
+            add_header "X-Otel-Parent-Id" $otel_parent_id;
+            add_header "X-Otel-Parent-Sampled" $otel_parent_sampled;
+            return 204;
+        }
+
+        location /context-ignore {
+            proxy_pass http://127.0.0.1:18080/notrace;
+        }
+
+        location /context-extract {
+            otel_trace_context extract;
+            proxy_pass http://127.0.0.1:18080/notrace;
+        }
+
+        location /context-inject {
+            otel_trace_context inject;
+            proxy_pass http://127.0.0.1:18080/notrace;
+        }
+
+        location /context-propagate {
+            otel_trace_context propagate;
+            proxy_pass http://127.0.0.1:18080/notrace;
+        }
+
+        location /notrace {
+            otel_trace off;
+            add_header "X-Otel-Trace-Id" $otel_trace_id;
+            add_header "X-Otel-Span-Id" $otel_span_id;
+            add_header "X-Otel-Parent-Id" $otel_parent_id;
+            add_header "X-Otel-Parent-Sampled" $otel_parent_sampled;
+            add_header "X-Otel-Traceparent" $http_traceparent;
+            add_header "X-Otel-Tracestate" $http_tracestate;
+            return 204;
+        }
+    }
+}
+
+"""
+
+(trace_id, span_id) = ("0af7651916cd43dd8448eb211c80319c", "b9c7c989f97918e1")
+context = {
+    "Traceparent": f"00-{trace_id}-{span_id}-01",
+    "Tracestate": "congo=ucfJifl5GOE,rojo=00f067aa0ba902b7",
+}
+
+
+def decode_id(span, value):
+    if value in ["trace_id", "span_id"]:
+        return binascii.hexlify(getattr(span, value)).decode("utf-8")
+    return value
+
+
+def get_attr(span, name):
+    for value in (a.value for a in span.attributes if a.key == name):
+        return getattr(value, value.WhichOneof("value"))
+
+
+def get_http09(url):
+    def do_get(sock, path):
+        http_send = f"GET {path}\n".encode()
+        sock.sendall(http_send)
+        http_recv = sock.recv(1024)
+        return http_recv.decode("utf-8")
+
+    p = urllib.parse.urlparse(url)
+    with socket.create_connection((p.hostname, p.port)) as sock:
+        recv = do_get(sock, p.path)
+    return recv
+
+
+@pytest.fixture
+def client(nginx):
+    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+    with niquests.Session(multiplexed=True) as s:
+        yield s
+
+
+def test_response_http09(trace_service, nginx):
+    assert get_http09("http://127.0.0.1:18080/ok") == "OK"
+
+    span = trace_service.get_span()
+
+    assert span.name == "/ok"
```

Comment:
As was already noted, there is no need to involve urllib for the sake of this test:

```suggestion
def test_http09(trace_service, nginx):
    def get_http09(host, port, path):
        with socket.create_connection((host, port)) as sock:
            sock.sendall(f"GET {path}\n".encode())
            return sock.recv(1024).decode("utf-8")

    assert get_http09("127.0.0.1", 18080, "/ok") == "OK"

    span = trace_service.get_span()

    assert span.name == "/ok"
```

Reply:
Applied.